src/snowflake/snowpark/_internal/type_utils.py

```diff
@@ -834,3 +835,10 @@ def type_string_to_type_object(type_str: str) -> DataType:
 ColumnOrSqlExpr = Union["snowflake.snowpark.column.Column", str]
 LiteralType = Union[VALID_PYTHON_TYPES_FOR_LITERAL_VALUE]
 ColumnOrLiteral = Union["snowflake.snowpark.column.Column", LiteralType]
+
+class CopyOptions(TypedDict, total=False):
```

Comment:
`TypedDict` and total arguments are supported starting with Python 3.8.

https://docs.python.org/3.8/library/typing.html?highlight=typeddict#typing.TypedDict